### DISCUSSION TITLE
Analysis Support in R

### DISCUSSION THREAD

**Discussion by mollywestbrook** [talmolab/sleap#1284] [General]
Hello Sleapers! While most of the analysis infrastructure of sleap is based around python, I was curious if anybody has done any model validation in R. I have had no trouble accessing the analysis .h5 file in R (for those in my situation, I have found the rhdf5 library to be quite helpful) and have been working with it fine, but many of the model validation files are in file formats I am unfamiliar with. I was checking out the validation notebooks on the website, but it seems like most of the validation is done through the help(sleap.load_metrics) command in python. Is there a good way to access the evaulation files in R? Thank you for any guidance you can give.

**Comment by phragmosis** [comment]
Hello Molly,

Sorry to be a pain, but is there any chance you could share the process/code of how you're converting the tracked coordinates from the h5 files into a csv for subsequent analysis in R? I have a student struggling with this at the moment, and I am completely unfamiliar with Python so can't help them!

Thanks so much.
Rob

**Reply by talmo** [reply]
Hi @phragmosis,

[Here's a Colab notebook](https://colab.research.google.com/drive/13Mt5we4keXRbMwJPXSjEogVucrqh81GF?usp=sharing) that can convert the H5 coordinates into a CSV that might be easier to work with in R.

Let us know (preferably in a new thread) if you have any questions!

Cheers,

Talmo

**Reply by mollywestbrook** [reply]
@phragmosis, 

The colab to convert to the deep lab cut (dlc) style csv is very handy. If you would like it, I have uploaded all of my preliminary sleap code to a repo. [This ](https://github.com/mollywestbrook/mollys-sleap-code/blob/main/sleaporganization.r) is the file I use to organize my h5 files (lmk if link isn't working). 

The final df here will not be like dlc style. I typically use tidyverse style coding in my r analysis, so I explicitly wanted long-form data. If your student codes like this, this shouldn't be a problem, but if you prefer to have something like dlc, use the colab notebook above and reorganize as they see fit. Hopefully between the two you should be set to do some preliminary analysis! Good luck! 

-- Molly

**Comment by talmo** [comment]
Hi @mollywestbrook,

Apologies for the delay! This is a bit tricky in R, but doable.

If you look at the source code for `sleap.load_metrics`, you'll note that it's basically just calling `np.load()` to read the file:
https://github.com/talmolab/sleap/blob/d05e2bf524cdb9813926a841886725644f00a218/sleap/nn/evals.py#L741-L778

I'm not super familiar with R, but asking ChatGPT for a function that can read NPZ files natively, I get this which looks to be about right:

> To read a numpy-formatted NPZ file in R, you can make use of the `reticulate` package, which provides a comprehensive set of tools for interoperability between Python and R. The `reticulate` package includes access to Python modules, classes, and functions. 
> 
> However, remember that to use this function, you need to have Python and numpy installed in your environment. 
> 
> Here is an example of how you can implement such a function:
> 
> ```R
> # Install the 'reticulate' package if you haven't already
> if (!require(reticulate)) {
>   install.packages('reticulate')
> }
> 
> # Load the 'reticulate' package
> library(reticulate)
> 
> # Import numpy in R through 'reticulate'
> np <- import("numpy")
> 
> read_npz <- function(file_path) {
>   # Load the numpy file
>   npzfile <- np$load(file_path, allow_pickle=TRUE)
> 
>   # List the keys in the npz file
>   keys <- npzfile$files
> 
>   # Create a named list to store the numpy arrays
>   numpy_arrays <- list()
> 
>   # Loop over the keys and add the numpy arrays to the list
>   for(key in keys){
>     numpy_arrays[[key]] <- npzfile[[key]]
>   }
> 
>   # Return the list of numpy arrays
>   return(numpy_arrays)
> }
> 
> # Use the function
> npz_file_path <- "/path/to/your/file.npz"
> numpy_arrays <- read_npz(npz_file_path)
> ```
> 
> This function `read_npz` takes the path to the NPZ file as input and returns a list of numpy arrays. The names of the list elements correspond to the keys in the NPZ file.
> 
> Please replace `"/path/to/your/file.npz"` with the actual path to your NPZ file.
> 
> Keep in mind that `reticulate` automatically converts between R and Python objects when they are passed from one language to the other. However, the conversion might not always work as expected, especially for more complex data types. It's important to check the resulting R object to make sure it has been converted correctly.

You may also want to check the [docs for reticulate](https://rstudio.github.io/reticulate/) for more reference on getting it set up.

Honestly though, it may be easier to just write a quick Python script based on our [evaluation notebook](https://sleap.ai/notebooks/Model_evaluation.html) that pulls out the specific metrics you want and saves them to CSV or another format a bit more friendly to R.

Let us know if you have any questions!

Cheers,

Talmo

**Reply by mollywestbrook** [reply]
Talmo,

Thank you so much for the assistance. I think the second option works better for me, to just write up a simple python script to calculate the error metrics for me ([this ](https://github.com/mollywestbrook/mollys-sleap-code/blob/main/sleaperroranalysis_python.py)is what I've got and it has been serving me well so far). I will probably end up trying out the r code at some point and will report back with how it works, but I think for now working with the python as needed will be the route going forward.
